app/models/user.rb

```diff
@@ -17,24 +17,47 @@ class User < ApplicationRecord
          :recoverable, :rememberable, :validatable,
          :doorkeeper, :omniauthable, omniauth_providers: [:google_oauth2]
 
+  enum provider: { email: 'email', google_oauth2: 'google_oauth2' }
+
+  validates :provider, inclusion: { in: providers.keys }
+
   class << self
-    def from_omniauth(auth)
-      User.find_or_create_by(provider: auth.provider, uid: auth.uid, email: auth.info.email) do |user|
-        set_user_attributes(user, auth)
-      end
+    def from_email(email, password)
+      user = find_or_initialize_user(
+        email: email,
+        password: password,
+        provider: User.providers[:email]
+      )
+
+      user.persisted? ? user.errors.add(:email, :taken) : user.save
+
+      user
     end
 
-    private
+    def from_omniauth(auth)
+      user = find_or_initialize_user(
+        email: auth.info.email,
+        password: Devise.friendly_token[0, 20],
+        provider: auth.provider,
+        auth: auth
+      )
 
-    def set_user_attributes(user, auth)
-      user.provider = auth.provider
-      user.uid = auth.uid
+      user.provider == User.providers[:email] ? user.errors.add(:email, :taken) : user.persisted? || user.save
```

Comment:
Do we need to check the `email` providers in this case? You are no longer using the `omniauth-identity` gem, right?

Reply:
Yes, we do. 🙏 

Let's say the user clicks the google sign-in button:

- If a user already had a `google_oauth2` account, then simply return the same user (no error)
- If a user already had an `email` account, then return an error that the e-mail already has been taken

Reply:
The code is too complex now 🙈 Ternary operator and another `||` operand 🤯 You need to find a simpler way to handle such complex logic.